### ISSUE TITLE
Support HTTP upgrades in MeshConfig.ProxyConfig

### ISSUE THREAD

**Issue by AnirbanNandi** [istio/istio#52651]
**Describe the feature request**

Add a new field to [MeshConfig.ProxyConfig](https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig) (say, `upgradeConfig`) that configures the Envoy proxy to perform HTTP upgrades. Under the hood, this could be something as simple as configuring the HTTP Connection Manager in Envoy, like, 
```
filters = append(filters, &listener.Filter{
   Name:       wellknown.HTTPConnectionManager,
   ConfigType: &listener.Filter_TypedConfig{TypedConfig: protoconv.MessageToAny(&hcm.HttpConnectionManager{
      ...
      UpgradeConfigs:   []*hcm.HttpConnectionManager_UpgradeConfig{
         &hcm.HttpConnectionManager_UpgradeConfig{UpgradeType: <protocol_name>/<version>},
      },
      ...
   })},
}
```
or it could be a bit more involved as well (I haven't assessed the tradeoffs yet).

**Additional context**

We use a type of k8s cluster where the client doesn't talk to the API server directly but instead the request passes through an ingress gateway which forwards it. As a result of this flow, `kubectl port-forward` and `exec` commands fail  as they rely on SPDY. As per [KEP 4006](https://github.com/kubernetes/enhancements/issues/4006), the transition from SPDY to WebSockets will be stable in 1.32 and it will be a long time before we can upgrade our clusters to that version. This FR is intended to be a general case of this HTTP->SPDY upgrade.

**Describe alternatives you've considered**

I do realise that we could just use the HCM EnvoyFilter to achieve the same thing and although they are very powerful, we generally tend to stay away from EnvoyFilters because of significant security and performance risks due to arbitrary changes.

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Dual Stack
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[X] Multi Cluster
[ ] Virtual Machine
[X] Multi Control Plane

**Comment by AnirbanNandi** [comment]
@keithmattix @ramaraochavali could you PTAL at this FR and share your thoughts?

cc @kyessenov

**Comment by ramaraochavali** [comment]
If we add it to Mesh config or Proxy config - how will we know for which listeners this should be applied? And in your example what would be the upgrade config look like?

**Comment by keithmattix** [comment]
The use-case seems compelling, but I echo @ramaraochavali's question re: the API. Feels like this would need to apply to DR or ServiceEntry or something

**Comment by AnirbanNandi** [comment]
Re @ramaraochavali's queries: 

> If we add it to Mesh config or Proxy config - how will we know for which listeners this should be applied? 

I was thinking we could combine the workload selector in ProxyConfig with a `port` field in upgrade config. So, maybe something like
```
apiVersion: networking.istio.io/v1beta1
kind: ProxyConfig
metadata:
  name: connection-upgrade-proxyconfig
  namespace: example-namespace
spec:
  selector:
    matchLabels:
      example-workload-label: example-workload-value
  upgradeConfig:
    port: PORT_NUM
    upgradeType: "protocol_name/version"
``` 

> what would be the upgrade config look like?

I assume you're asking about the field that would be exposed in ProxyConfig, in which case, I was thinking of having only the upgrade type (protocol name and version) initially. I realise that HCM provides more [customisations](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-upgradeconfig) but I guess I'm more focused on the upgrade itself right now. I suppose this can be extended later by introducing new fields or using existing fields in ProxyConfig to achieve envoy-filter-chain-like customisation (e.g. `forwardedClientCert` from `ProxyConfig.ProxyHeaders` can be used instead of [extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.ForwardClientCertDetails](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#enum-extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-forwardclientcertdetails))

-------

Re @keithmattix's suggestion on using "DR or ServiceEntry or something", I think maybe the API should be exposed in a CR such that the setting can be applied on all (relevant) resources in the mesh without needing any extra steps. e.g, it should be possible to do connection upgrades on a service inside the mesh without having to add a ServiceEntry CR. Similarly, the API should be able to upgrade connections to workloads as well as ingress gateways. Wdyt?

PS - I'm not strongly fixated on exposing this capability via ProxyConfig :) but to me, it seems like the most relevant CR that can mutate all types of resources in the mesh. Open to other suggestions/viewpoints, of course.

**Comment by keithmattix** [comment]
Wait: clarifying question - are you trying to change how a sidecar talks to the k8s api server or how istiod talks to k8s api server? If the former, then a service entry + DR for the k8s API server sounds like the right layer of abstraction since it's just another service in the mesh

**Comment by AnirbanNandi** [comment]
> Wait: clarifying question - are you trying to change how a sidecar talks to the k8s api server or how istiod talks to k8s api server?

Sidecar to k8s api server

So IIUC (and please correct me if I'm wrong), you're suggestion is to do something like
```
apiVersion: networking.istio.io/v1
kind: DestinationRule
metadata:
  name: ingress-gateway-upgrade
spec:
  host: istio-ingressgateway.istio-system.svc.cluster.local
  trafficPolicy: 
    portLevelSettings:
    - port:
        number: PORT_NUM
      upgradeConfig:
        upgradeType: "protocol_name/version"
```

If that is what you had in mind, then this makes sense but I have two questions
1. Why do we need ServiceEntry in addition to the DestinationRule? Is it to extend the connection upgrade capability to mesh-internal services that not part of the service registry (e.g., VMs running Envoy that are talking to K8s services)?
2. Can I ask why this is a more preferable abstraction over proxy config?

**Comment by keithmattix** [comment]
Yeah something like that should work since you're sending your apiserver requests through a gateway. You're right you don't need a ServiceEntry; I was confused about the scenario. The reason this is preferred over ProxyConfig is because it's more generic, allowing upgrades for any host, port, etc. rather than proxy config which typically describes the behavior and attributes of the proxy itself rather than specific routing rules or transport settings.

**Comment by AnirbanNandi** [comment]
I see. Yeah that makes sense. Could you assign this to me? I can get started with the implementation with a minimal upgrade config.

**Comment by howardjohn** [comment]
Maybe a dumb question, but do we actually want envoy to proxy the upgrade or just any unknown upgrade through? ref https://github.com/istio/istio/issues/53239

**Comment by Stono** [comment]
Ignore unknown please! 
I'm legitimately worried about raft of users who are going to run into the issue from the other thread as they start upgrading libraries (we're quite on top of upgrades).  It's worse because it "works locally but not when deployed" - the worst type of problem.

**Comment by kyessenov** [comment]
I think it's fine to treat TLS and SPDY or any unknown as websocket-like plain TCP traffic on inbound. Does the authorization policy work the same for upgraded and non-upgraded traffic inbound? I think it doesn't matter if you do HTTP + upgrade vs direct unlike CONNECT where you can bypass port policy by upgrading from another port, but this is worth thinking about.
